File: AkkaSTMExample/src/main/java/org/akka/essentials/stm/pingpong/example2/PingPong.java
package org.akka.essentials.stm.pingpong.example2;

import scala.concurrent.stm.Ref;

public class PingPong {

	//updates to Ref.View are synchronous
	Ref.View<String> whoseTurn;

	public PingPong(Ref.View<String> player) {
		whoseTurn = player;
	}

	public boolean hit(final String opponent) {
		final String x = Thread.currentThread().getName();

		if (whoseTurn.get() == "") {
			whoseTurn.set(x);
			return true;
		}

		if (x.compareTo(whoseTurn.get()) == 0) {
			System.out.println("PING! (" + x + ")");
			whoseTurn.set(opponent);
		} else {
			try {
				long t1 = System.currentTimeMillis();
				wait(2500);
				if ((System.currentTimeMillis() - t1) > 2500) {
					System.out.println("****** TIMEOUT! " + x
							+ " is waiting for " + whoseTurn + " to play.");
				}
			} catch (Exception e) {
			}
		}
		if (whoseTurn.get().compareTo("DONE") == 0)
			return false;

		if (opponent.compareTo("DONE") == 0) {
			whoseTurn.set(opponent);
			return false;
		}
		return true; // keep playing.
	}
}

File: AkkaSTMExample/src/main/scala/org/akka/essentials/stm/pingpong/example2/PingPong.scala
package org.akka.essentials.stm.pingpong.example2

import scala.concurrent.stm.Ref

class PingPong(whoseTurn: Ref[String]) {

  def hit(opponent: String): Boolean = {

    val x: String = Thread.currentThread().getName()

    if (whoseTurn.single.get == "") {
      whoseTurn.single.set(x)
      return true
    } else if (whoseTurn.single.get.compareTo(x) == 0) {
      println("PING! (" + x + ")")
      whoseTurn.single.set(opponent)
      return true
    } else {
      try {
        var t1 = System.currentTimeMillis()
        wait(2500)
        if ((System.currentTimeMillis() - t1) > 2500) {
          println("****** TIMEOUT! " + x
            + " is waiting for " + whoseTurn + " to play.")
        }
      } catch {
        case _ =>
        // eat the exception
      }
    }
    if (opponent.compareTo("DONE") == 0) {
      whoseTurn.single.set(opponent)
      return false
    }
    if (whoseTurn.single.get.compareTo("DONE") == 0){
      return false
    }
    return true // keep playing.
  }
}